Mathlib/CategoryTheory/Sites/Equivalence.lean

```diff
@@ -247,4 +382,18 @@ instance hasColimitsEssentiallySmallSite
   Adjunction.has_colimits_of_equivalence ((equivSmallModel C).sheafCongr J
     ((equivSmallModel C).locallyCoverDense J).inducedTopology A).functor
 
+open ConcreteCategory
+
+variable [ConcreteCategory.{w} A]
+variable [HasFunctorialSurjectiveInjectiveFactorization A]
+  [((equivSmallModel C).locallyCoverDense J).inducedTopology.WEqualsLocallyBijective A]
+variable [((equivSmallModel C).locallyCoverDense J).inducedTopology.HasSheafCompose (forget A)]
+variable [Balanced (Sheaf ((equivSmallModel C).locallyCoverDense J).inducedTopology A)]
+variable {F G : Sheaf J A} (f : F ⟶ G)
+
+lemma Sheaf.isLocallySurjective_iff_epi_of_site_essentiallySmall :
```

Comment:
I would think that a better design would be to transport the `WEqualsLocallyBijective`, `HasSheafCompose` and `Balanced` instances via an equivalence. In the applications to specific sites, once these instances have been obtained, we may use `isLocallySurjective_iff_epi'`. This would also allow to use these instances for other purposes that just comparing locally surjective morphisms and epimorphisms. (This can be done in separate PRs.)

Reply:
You're right, I took the lazy approach.. I'll do it properly :)

Reply:
I know it is a little bit painful work, but I believe we have most of the ingredients.

Reply:
It wasn't too bad. However, using this approach one has to assume the existence of certain limits in the target category, so the "lazy" approach is slightly more general. I don't know if it's worth keeping that lemma though (without the limit assumption), because this only applies to concrete categories anyway.

Reply:
In order to proceed next with this material, I would suggest that for this PR, you keep only the general lemmas `isLocallySurjective/Injective_{/of/iff}_whisker` under quite general conditions (cocontinuous, etc) -- and these should be moved to another (maybe new) file. You may also keep the verification that equivalences are co/continuous in this PR, but the rest should very much be in differents PRs (specifically https://github.com/leanprover-community/mathlib4/pull/13486/files#diff-fd03ab2b884c2d15f411302648dd33c4f1e0db0c79f4ccaf8824354f59815bffR342-R378 and https://github.com/leanprover-community/mathlib4/pull/13486/files#diff-fd03ab2b884c2d15f411302648dd33c4f1e0db0c79f4ccaf8824354f59815bffR417-R446); actually, part of this material, especially the first chunk, will eventually be uncessary: suppose I have a site `(C,J)` which is big, and an equivalent one `(D,K)` which is small, we transport the necessary instances from `(D,K)` to `(C,J)` one by one, and then we may apply `isLocallySurjective_iff_epi'` in `(C,J)`. I think this is what you intend to do for the light profinite site, and this is what I intend to do in the future when the categories `C` and `D` are not necessarily equivalent, but at least they would have equivalent categories of sheaves (e.g. `C` is a category of schemes, and `D` is a (small) full subcategory of affines schemes).

Then, after this first PR is merged, we may proceed with transport lemmas, separately for `WEqualsLocallyBijective` (for an equivalence, I do not think we need more `HasLimitsOfShape` assumptions), `HasSheafCompose` and `Balanced`. It will be easier to work on this and review it as separate PRs.